chromium_src/chrome/browser/ui/webui/chrome_web_ui_configs.cc

```diff
@@ -19,20 +19,41 @@
 #include "brave/browser/ui/webui/brave_shields/cookie_list_opt_in_ui.h"
 #include "brave/browser/ui/webui/brave_shields/shields_panel_ui.h"
 #include "brave/browser/ui/webui/brave_wallet/wallet_panel_ui.h"
+#include "brave/browser/ui/webui/private_new_tab_page/brave_private_new_tab_ui.h"
 #include "brave/browser/ui/webui/speedreader/speedreader_toolbar_ui.h"
 #include "brave/browser/ui/webui/webcompat_reporter/webcompat_reporter_ui.h"
 #endif  // !BUILDFLAG(IS_ANDROID)
 #include "brave/browser/ui/webui/brave_adblock_internals_ui.h"
 #include "brave/browser/ui/webui/brave_adblock_ui.h"
 
+namespace {
+
+#if !BUILDFLAG(IS_ANDROID)
+const GURL GetWebUIConfigURL(const char* scheme, const char* host) {
+  return GURL(base::StrCat({scheme, url::kStandardSchemeSeparator, host}));
+}
+#endif  // !BUILDFLAG(IS_ANDROID)
+
+void RemoveOverridenWebUIs(content::WebUIConfigMap& map) {
+#if !BUILDFLAG(IS_ANDROID)
+  // NewTabUIConfig (overriden with BravePrivateNewTabUIConfig)
+  map.RemoveConfig(
+      GetWebUIConfigURL(content::kChromeUIScheme, chrome::kChromeUINewTabHost));
+#endif  // !BUILDFLAG(IS_ANDROID)
+}
+
+}  // namespace
+
 void RegisterChromeWebUIConfigs() {
   RegisterChromeWebUIConfigs_ChromiumImpl();
 
   auto& map = content::WebUIConfigMap::GetInstance();
+  RemoveOverridenWebUIs(map);
```

Comment:
I guess it chooses between regular and private based on `IsWebUIEnabled`? That's definitely not obvious and I also don't understand how this comment https://github.com/brave/brave-core/pull/25662/commits/abf9e14fec288b9b96d81ff0d0e79240f9cf1ed8#diff-9fedf355253a1cbae9d10ffd28405a8d217ecd068f319c5a08cab609b0cd398cR39 relates to that

Reply:
is there a test that covers this functionality (choosing regular vs private ntp)?

Reply:
I think we need some comments for `IsWebUIEnabled` and also https://github.com/brave/brave-core/pull/25662/commits/abf9e14fec288b9b96d81ff0d0e79240f9cf1ed8#diff-739b827f200537078f68db9cca1601599dac34e7b4077573ba198a925365df27R137

but why is `BraveNewTabUI` still in `NewWebUI`, but `BravePrivateNewTabUIConfig` is here in `RegisterChromeWebUIConfigs`?

Reply:
I put the function there to make it more generic in case we have some Android WebUIs that we also override and need the upstream implementations removed. 

I think the choice between the regular and private in the upstream's code happens elsewhere based on the [comment in chrome_web_ui_controller_factory.cc](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ui/webui/chrome_web_ui_controller_factory.cc;l=275): 
```
  if (url.host_piece() == chrome::kChromeUINewTabHost) {
    // The URL chrome://newtab/ can be either a virtual or a real URL,
    // depending on the context. In this case, it is always a real URL that
    // points to the New Tab page for the incognito profile only. For other
    // profile types, this URL must already be redirected to a different URL
    // that matches the profile type.
    //
    // Returning NewWebUI<NewTabUI> for the wrong profile type will lead to
    // crash in NTPResourceCache::GetNewTabHTML (Check: false), so here we add
    // a sanity check to prevent further crashes.
    //
    // The switch statement below must be consistent with the code in
    // NTPResourceCache::GetNewTabHTML!
    switch (NTPResourceCache::GetWindowType(profile)) {
      case NTPResourceCache::NORMAL:
        LOG(ERROR) << "Requested load of chrome://newtab/ for incorrect "
                      "profile type.";
        // TODO(crbug.com/40244589): Add DumpWithoutCrashing() here.
        return nullptr;
      case NTPResourceCache::INCOGNITO:
        [[fallthrough]];
      case NTPResourceCache::GUEST:
        [[fallthrough]];
      case NTPResourceCache::NON_PRIMARY_OTR:
        return &NewWebUI<NewTabUI>;
    }
```
We are getting around that assumption in the upstream code elsewhere because `NewWebUI` in `brave/browser/ui/webui/brave_web_ui_controller_factory.cc` gets called for a regular profile. Initially it was via https://github.com/brave/brave-core/pull/20/files#diff-594631d3a40b383b831cd6c6c401ff278c4b259ec7f1b321808e868ccfb06f2bR15, but I am not sure what this changed to in the current code.

Reply:
Added comments in https://github.com/brave/brave-core/pull/25662/commits/63f646a7d48654144c9f28b2c58d93fe4effacd8